### PR TITLE
Encapsulate program-level options

### PR DESCRIPTION
**Pull request by nwatson22**
Adds `ProveOptions` dataclass for storing the settings that don't change per `kontrol prove` execution

### CODE REVIEW

File: src/kontrol/__main__.py [hunk 1]
```diff
@@ -528,6 +529,13 @@ def _parse_test_version_tuple(value: str) -> tuple[str, int | None]:
         action='store_true',
         help='Include the contract constructor in the test execution.',
     )
+    prove_args.add_argument(
+        '--fail-fast',
+        dest='fail_fast',
+        default=False,
+        action='store_true',
+        help='Stop execution on other branches if failing node is detected.',
```
Comment (palinatolmach) [inline src/kontrol/__main__.py:537]:
```suggestion
        help='Stop execution on other branches if a failing node is detected.',
```

File: src/kontrol/options.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    from collections.abc import Iterable
+    from pathlib import Path
+
+    from pyk.utils import BugReport
+
+    from .foundry import Foundry
+
+
+@dataclass
+class GlobalOptions:
```
Comment (palinatolmach) [inline src/kontrol/options.py:16]: I'm wondering if the class name should reflect the fact that these settings hold throughout each `prove` run, e.g., `ProveOptions`? I'm not sure, so please feel free to ignore that.

File: src/kontrol/prove.py [hunk 3]
```diff
@@ -293,15 +280,12 @@ def init_and_run_proof(test: FoundryTest) -> tuple[bool, list[str] | None]:
 
 
 def method_to_apr_proof(
-    foundry: Foundry,
     test: FoundryTest,
     kcfg_explore: KCFGExplore,
-    simplify_init: bool = True,
-    bmc_depth: int | None = None,
-    run_constructor: bool = False,
+    options: GlobalOptions,
```
[the remark this answers is not in the document]
Comment (nwatson22) [reply src/kontrol/prove.py]: Reverted to using individual arguments in this function.

File: src/kontrol/prove.py [hunk 4]
```diff
@@ -108,28 +110,35 @@ def foundry_prove(
     for test in constructor_tests:
         test.method.update_digest(foundry.digest_file)
 
+    llvm_definition_dir = foundry.llvm_library if use_booster else None
+
+    options = GlobalOptions(
```
[the remark this answers is not in the document]
Comment (nwatson22) [reply src/kontrol/prove.py]: Moved into `exec_prove` and removed `foundry` and `llvm_definition_dir` fields.

File: src/kontrol/__main__.py (hunk 1 above)
Comment (nwatson22) [reply src/kontrol/__main__.py]: Fixed.

File: src/kontrol/options.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nwatson22) [reply src/kontrol/options.py]: Done.

Comment (nwatson22) [reply src/kontrol/options.py]: Renamed to `ProveOptions`.

File: src/kontrol/prove.py [hunk 5]
```diff
@@ -107,30 +89,16 @@ def foundry_prove(
     for test in constructor_tests:
         test.method.update_digest(foundry.digest_file)
 
+    #      llvm_definition_dir = foundry.llvm_library if options.use_booster else None
+
```
Comment (nwatson22) [inline src/kontrol/prove.py:93]:
```suggestion
```

File: src/kontrol/options.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+    from pyk.utils import BugReport
+
+
+@dataclass(frozen=True)
+class ProveOptions:
+    auto_abstract_gas: bool
+    bug_report: BugReport | None
+    use_booster: bool
+    kore_rpc_command: str | Iterable[str] | None
```
Comment (tothtamas28) [inline src/kontrol/options.py:17]: Now that you have an `__init__`, you can strengthen the types, e.g.:

```suggestion
    kore_rpc_command: tuple[str, ...]
```

File: src/kontrol/options.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        workers: int = 1,
+        counterexample_info: bool = False,
+        max_iterations: int | None = None,
+        run_constructor: bool = False,
+        fail_fast: bool = True,
+        reinit: bool = False,
+        port: int | None = None,
+    ) -> None:
+        if kore_rpc_command is None:
+            kore_rpc_command = ('kore-rpc-booster',) if use_booster else ('kore-rpc',)
```
Comment (tothtamas28) [inline src/kontrol/options.py:58]: To ensure declared types are as straightforward as possible, you can for example:

```suggestion
        if kore_rpc_command is None:
            kore_rpc_command = ('kore-rpc-booster',) if use_booster else ('kore-rpc',)
        elif isinstance(kore_rpc_command, str):
            kore_rpc_command = (kore_rpc_command,)
        else:
            kore_rpc_command = tuple(kore_rpc_command)
```

File: src/kontrol/prove.py [hunk 8]
```diff
@@ -349,7 +297,7 @@ def _load_constructor_proof(foundry: Foundry, contract: Contract) -> APRProof:
     return apr_proof
 
 
-def _method_to_initialized_cfg(
+def method_to_initialized_cfg(
```
Comment (tothtamas28) [inline src/kontrol/prove.py:300]: Is this change intentional?

**Review by palinatolmach — APPROVED** [review]

File: src/tests/integration/test_foundry_prove.py [hunk 9]
```diff
@@ -169,8 +172,10 @@ def test_foundry_fail(
     prove_res = foundry_prove(
         foundry_root,
         tests=[(test_id, None)],
-        counterexample_info=True,
-        port=server.port,
+        options=ProveOptions(
+            counterexample_info=True,
```
Comment (tothtamas28) [inline src/tests/integration/test_foundry_prove.py:176]: We pass the bug report here too.

File: src/tests/integration/test_foundry_prove.py [hunk 10]
```diff
@@ -412,10 +433,12 @@ def test_foundry_init_code(test: str, foundry_root: Path, use_booster: bool) ->
     prove_res = foundry_prove(
         foundry_root,
         tests=[(test, None)],
-        smt_timeout=300,
-        smt_retry_limit=10,
-        use_booster=use_booster,
-        run_constructor=True,
+        options=ProveOptions(
```
Comment (tothtamas28) [inline src/tests/integration/test_foundry_prove.py:436]: And also here.

File: src/tests/profiling/test_foundry_prove.py [hunk 11]
```diff
@@ -36,10 +37,12 @@ def test_foundy_prove(profile: Profiler, use_booster: bool, tmp_path: Path) -> N
         foundry_prove(
             foundry_root,
             tests=[('AssertTest.test_revert_branch', None)],
-            smt_timeout=300,
-            smt_retry_limit=10,
-            counterexample_info=True,
-            use_booster=use_booster,
+            options=ProveOptions(
```
Comment (tothtamas28) [inline src/tests/profiling/test_foundry_prove.py:40]: And here.

File: src/kontrol/options.py (hunk 6 above)
Comment (nwatson22) [reply src/kontrol/options.py]: Fixed.

File: src/kontrol/options.py (hunk 7 above)
Comment (nwatson22) [reply src/kontrol/options.py:58]: I added this. For the other fields, I left the types as-is since for the ones that can be `None` this has a specific meaning in the rest of the code.

File: src/kontrol/prove.py (hunk 8 above)
Comment (nwatson22) [reply src/kontrol/prove.py]: I reverted it. This got introduced in a manual merge.

File: src/tests/integration/test_foundry_prove.py (hunk 9 above)
Comment (nwatson22) [reply src/tests/integration/test_foundry_prove.py:181]: Done.

File: src/tests/integration/test_foundry_prove.py (hunk 10 above)
Comment (nwatson22) [reply src/tests/integration/test_foundry_prove.py:442]: Done.

File: src/tests/profiling/test_foundry_prove.py (hunk 11 above)
Comment (nwatson22) [reply src/tests/profiling/test_foundry_prove.py:41]: Done.

File: src/kontrol/options.py [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+    counterexample_info: bool
+    max_iterations: int | None
+    run_constructor: bool
+    fail_fast: bool
+    reinit: bool
+    port: int | None
+
+    def __init__(
+        self,
+        auto_abstract_gas: bool = False,
```
Comment (tothtamas28) [inline src/kontrol/options.py:37]:
```suggestion
    def __init__(
        self,
        *,
        auto_abstract_gas: bool = False,
```

**Review by tothtamas28 — APPROVED** [review]
Looks good!
